src/Sarif.Multitool/Rules/ContextRegionRequiresRegion.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class ContextRegionRequiresRegion : SarifValidationSkimmerBase
+    {
+        private readonly Message _fullDescription = new Message
+        {
+            Text = RuleResources.SARIF1016_ContextRegionRequiresRegion
+        };
+
+        public override Message FullDescription => _fullDescription;
+
+        public override ResultLevel DefaultLevel => ResultLevel.Error;
+
+        /// <summary>
+        /// SARIF1016
+        /// </summary>
+        public override string Id => RuleId.ContextRegionRequiresRegion;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1016_Default)
+        };
+
+        protected override void Analyze(PhysicalLocation physicalLocation, string physicalLocationPointer)
+        {
+            if (physicalLocation.ContextRegion != null && physicalLocation.Region == null)
```

Comment:
>physicalLocation.ContextRegion [](start = 16, length = 30)

What if someone provides an entirely empty region? A region has no required items. What if someone populated the context region with line/other information, but the Region object only includes a 'Message' string? Etc.

Reply:
The schema doesn't declare any required properties, but a `region` object with no properties is invalid, because:

- §3.25.5: When a `region` object represents a text region specified by line/column properties, it **SHALL** contain a property named `startLine`...
- §3.25.9: When a `region` object represents a text region specified by offset/length properties, it **SHALL** contain a property named `charOffset`...
- §3.25.11: When a `region` object represents a binary region, it **SHALL** contain a property named `byteOffset`...

I filed #1143: "Add validation rule: at least one of the region "start" properties must be present."

---
In reply to: [238499344](https://github.com/Microsoft/sarif-sdk/pull/1142#discussion_r238499344) [](ancestors = 238499344)

Reply:
You should satisfy that work item by updating the schema itself, rather than authoring a validator rule. An 'anyOf' rule should cover this, e.g.:

  "anyOf": [
    {"required" : ["startLine"]},
    {"required" : ["charOffset"]},
    {"required" : ["byteOffset"]}
  ]

IOW, we should favor schema-driven solutions wherever possible. Of course, this may prompt us to author improved validation within jschema itself...

---
In reply to: [238504114](https://github.com/Microsoft/sarif-sdk/pull/1142#discussion_r238504114) [](ancestors = 238504114,238499344)